### ISSUE TITLE
AttributeError in webelem.rect_on_view

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#1641]
Crash report on v0.7.0, when hinting on voidlinux.eu with `content -> allow-javascript = false`, Qt 5.6.1:

```
04:41:05 DEBUG    commands   command:run:488 command called: hint
04:41:05 DEBUG    commands   command:run:503 Calling qutebrowser.browser.hints.HintManager.start(<qutebrowser.browser.hints.HintManager object at 0x7fe9e5848dc8>, False, <Group.all: 1>, <Target.normal: 1>, win_id=0)
04:41:05 DEBUG    hints      hints:_init_elements:701 hints: f, wd, j, wk, d, ws, k, wl, s, wa, l, w;, a, wr, ;, wu, r, we, u, wi, e, ww, i, wo, wf, of, wj
04:41:05 ERROR    misc       crashsignal:exception_hook:211 Uncaught exception
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/qutebrowser/app.py", line 865, in eventFilter
    return handler(event)
  File "/usr/lib/python3.4/site-packages/qutebrowser/app.py", line 825, in _handle_key_event
    return man.eventFilter(event)
  File "/usr/lib/python3.4/site-packages/qutebrowser/keyinput/modeman.py", line 318, in eventFilter
    return self._eventFilter_keypress(event)
  File "/usr/lib/python3.4/site-packages/qutebrowser/keyinput/modeman.py", line 170, in _eventFilter_keypress
    handled = parser.handle(event)
  File "/usr/lib/python3.4/site-packages/qutebrowser/keyinput/basekeyparser.py", line 309, in handle
    match = self._handle_single_key(e)
  File "/usr/lib/python3.4/site-packages/qutebrowser/keyinput/modeparsers.py", line 76, in _handle_single_key
    match = super()._handle_single_key(e)
  File "/usr/lib/python3.4/site-packages/qutebrowser/keyinput/basekeyparser.py", line 197, in _handle_single_key
    self.execute(binding, self.Type.chain, count)
  File "/usr/lib/python3.4/site-packages/qutebrowser/keyinput/keyparser.py", line 44, in execute
    self._commandrunner.run(cmdstr, count)
  File "/usr/lib/python3.4/site-packages/qutebrowser/commands/runners.py", line 286, in run
    result.cmd.run(self._win_id, args)
  File "/usr/lib/python3.4/site-packages/qutebrowser/commands/command.py", line 504, in run
    self.handler(*posargs, **kwargs)
  File "/usr/lib/python3.4/site-packages/qutebrowser/browser/hints.py", line 854, in start
    self._init_elements()
  File "/usr/lib/python3.4/site-packages/qutebrowser/browser/hints.py", line 703, in _init_elements
    label = self._draw_label(e, string)
  File "/usr/lib/python3.4/site-packages/qutebrowser/browser/hints.py", line 417, in _draw_label
    self._set_style_properties(elem, label)
  File "/usr/lib/python3.4/site-packages/qutebrowser/browser/hints.py", line 375, in _set_style_properties
    self._set_style_position(elem, label)
  File "/usr/lib/python3.4/site-packages/qutebrowser/browser/hints.py", line 385, in _set_style_position
    rect = elem.rect_on_view(adjust_zoom=False, no_js=no_js)
  File "/usr/lib/python3.4/site-packages/qutebrowser/browser/webelem.py", line 178, in rect_on_view
    return rect_on_view(self._elem, **kwargs)
  File "/usr/lib/python3.4/site-packages/qutebrowser/browser/webelem.py", line 401, in rect_on_view
    for i in range(int(rects.get("length", 0))):
AttributeError: 'NoneType' object has no attribute 'get'
```

cc @lahwaacz

**Comment by lahwaacz** [comment]
That's strange, apparently the `elem.evaluateJavaScript()` method behaves inconsistently with respect to the `content -> allow-javascript` option on different Qt releases (or OSes?).

I could not reproduce this, but the fix might be
```diff
diff --git a/qutebrowser/browser/webkit/webelem.py b/qutebrowser/browser/webkit/webelem.py
index bb671b3..ea95dae 100644
--- a/qutebrowser/browser/webkit/webelem.py
+++ b/qutebrowser/browser/webkit/webelem.py
@@ -392,7 +392,11 @@ class WebElementWrapper(collections.abc.MutableMapping):
             text = utils.compact_text(self._elem.toOuterXml(), 500)
             log.hints.vdebug("Client rectangles of element '{}': {}".format(
                 text, rects))
-            for i in range(int(rects.get("length", 0))):
+            if isinstance(rects, dict):
+                length = int(rects.get("length", 0))
+            else:
+                length = 0
+            for i in range(length):
                 rect = rects[str(i)]
                 width = rect.get("width", 0)
                 height = rect.get("height", 0)
```

**Comment by The-Compiler** [comment]
I suspected that as well, but I have [a test](https://github.com/The-Compiler/qutebrowser/blob/master/tests/unit/browser/webkit/test_qt_javascript.py#L58-L64) which works everywhere...

I'll fix it by splitting `rect_on_view` into two methods.

**Comment by The-Compiler** [comment]
FWIW I just got a second crash report, with JS turned on, on Void Linux. I don't know the distribution for the original reporter ("Linux-4.5.2-aufs-sysresc-TM-B113-x86_64-with-libc"), but they were visiting voidlinux.eu...

**Comment by The-Compiler** [comment]
I just got another one on v0.8.2 on Void Linux. Seems like the fix didn't actually make it into v0.8.2, but it should now be fixed for v0.8.3 in a43aa52c81ee379031f0d189889e4b45ff94824f (and on `master` still via 56515321dd3670ea41abed39378cff1b7c984910).

**Comment by The-Compiler** [comment]
By the way, according to someone who did run into that bug, it affects people using musl libc (which often happen to be Void Linux users :wink:) because its stack size (80k) is too small for QtWebKit's javascript engine. glibc used 8 MB as default stack size.
